clarity/src/vm/callables.rs

```diff
@@ -300,3 +352,295 @@ impl FunctionIdentifier {
+                } else {
+                    &inner_types.1
+                },
+                data,
+            )?),
+        }),
+        (
+            TypeSignature::SequenceType(SequenceSubtype::ListType(list_type)),
+            Value::Sequence(SequenceData::List(ListData {
+                data,
+                type_signature,
+            })),
+        ) => {
+            let mut values = Vec::with_capacity(data.len());
+            for elem in data {
+                values.push(clarity2_implicit_cast(
+                    list_type.get_list_item_type(),
+                    elem,
+                )?);
+            }
+            let cast_list_type_data = ListTypeData::new_list(
+                list_type.get_list_item_type().clone(),
+                type_signature.get_max_len(),
+            )?;
+            Value::Sequence(SequenceData::List(ListData {
+                data: values,
+                type_signature: cast_list_type_data,
+            }))
+        }
+        (
+            TypeSignature::TupleType(tuple_type),
+            Value::Tuple(TupleData {
+                type_signature: _,
+                data_map,
+            }),
+        ) => {
+            let mut cast_data_map = BTreeMap::new();
+            for (name, field_value) in data_map {
+                let to_type = match tuple_type.get_type_map().get(name) {
+                    Some(ty) => ty,
+                    None => {
+                        // This should be unreachable if the type-checker has already run successfully
+                        return Err(
+                            CheckErrors::TypeValueError(type_sig.clone(), value.clone()).into()
+                        );
+                    }
+                };
+                cast_data_map.insert(name.clone(), clarity2_implicit_cast(to_type, field_value)?);
+            }
+            Value::Tuple(TupleData {
+                type_signature: tuple_type.clone(),
+                data_map: cast_data_map,
+            })
+        }
+        (
+            TypeSignature::CallableType(CallableSubtype::Trait(trait_identifier)),
+            Value::CallableContract(callable_data),
+        ) => Value::CallableContract(CallableData {
+            contract_identifier: callable_data.contract_identifier.clone(),
+            trait_identifier: Some(trait_identifier.clone()),
+        }),
+        // N.B. it seems like this should be illegal, since it is converting a
+        // principal to a callable trait, and only principal literals should be
+        // allowed to do that. The case that this is handling is when principal
+        // values are passed in from the initial contract-call, which by
+        // definition must be a literal. Other scenarios where a principal is
+        // passed will have been caught by the type checker. This could
+        // alternatively be checked with
+        // `FunctionType::check_args_by_allowing_trait_cast` before execution.
```

Comment:
Would this imply that this code can only be reached through the top-level call to `clarity2_implicit_cast()`?

Also, what about the case when a `TransactionPayload::ContractCall(..)` transaction gets mined and contains a trait concretization embedded in a compound type?  For example, suppose I send a transaction that contains a `(some 'SPxxx.foo)` argument value because the target function accepts `(optional <foo-trait>)`?  Is that case handled?

Reply:
No, this does not imply that it can only be reached by the top-level call. It also handles the recursive cases, such as the `(optional <foo-trait>)` that you described.

Reply:
Great, thanks!